lib/BackbeatConsumer.js

```diff
@@ -923,10 +921,20 @@ class BackbeatConsumer extends EventEmitter {
                 if (this._zookeeper) {
                     this._zookeeper.close();
                 }
-                if (this._consumer) {
+                if (this._consumer && this._consumer.isConnected()) {
                     this._consumer.unsubscribe();
-                    this._consumer.disconnect();
-                    this._consumer.on('disconnected', () => next());
+                    // Wait for partition unassign to complete before
+                    // disconnecting, the rebalance callback will handle
+                    // waiting for current jobs to complete as well as commit
+                    // the latest offsets
+                    this.once('unassign', status => {
+                        if (status === UNASSIGN_STATUS.TIMEOUT) {
```

Comment:
when this happens, can we/should we wait for `disconnected` ?
disconnect() will have been called, not sure if we will get the event...

Reply:
I added a test case for this.
If the event already got emitted before adding the event handler then we will never receive the event yes.